Paginated Help Command

YoussefExplicit opened the pull request:
- One collector
- Moved the function to the top and now it uses this.pages()
- Hope you like it Kyra.

YoussefExplicit committed:
Create help.js

YoussefExplicit committed:
Delete help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
fixed some stuff

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
updated a lot.

YoussefExplicit committed:
:)

YoussefExplicit committed:
fixed some minor mistakes

YoussefExplicit committed:
One reactor :)

YoussefExplicit committed:
Moved function to the top 

Moved the function to the top so you can't get the error for 'pages' was used before it was defined in eslint

YoussefExplicit committed:
Ability to use this.pages

T4LO5 commented:
Just going to leave this in here :- 
https://pastebin.com/L7mjcudQ

YoussefElshemi commented:
@T4LO5 Oh yh...

YoussefExplicit committed:
" -> '

kyranet commented on commands/system/help.js:
```diff
@@ -0,0 +1,191 @@
+      if (hm === "backward") {
+        if (page === 0) return;
+      }
+      await msg2.edit({
```
Should use the MessageEmbed constructor.

kyranet reviewed (changes requested):
I have a question for you, @YoussefElshemi, why did you open #64, #66 and now, #68, whenever you partially fix my requested changes? It's making my job much harder and also making me taking longer whilst I also lose my time. I have to remind you that this PR will not merge until I give my approval, and by closing and opening duplicated PRs, you're only getting me tired of re-reviewing the entire file, it's a disrespect and also goes against many conventions.

The same PR you open to fix something, the same PR we close once it's finished, please don't create any other PR to circumvent my requested changes.

I'm highly considering to close this PR (#68) and re-open the original one, #64.

A `CONTRIBUTING.md` file will be discussed by me with @YorkAARGH and the staff and pushed to this repository to avoid this issue in the future.

AndrewJWin commented:
@kyranet The reason for the closing/opening new PRs is my fault, it was due to a shared issue between myself and @YoussefElshemi, where commits would not show up, therefore the PR would be out of date. It was not meant as a disrespect to your requested changes, mainly as an acceptance and trying to amend them.

YoussefExplicit committed:
using messageembed constructor

AndrewJWin reviewed (changes requested):
Oi, Travis will hate this, your apostrophes are back. Change ' to "

YoussefExplicit committed:
oops

YoussefElshemi commented:
Done!

YorkAARGH commented:
You need to fix your conflicts

YoussefExplicit committed:
Merge branch 'Yukine-Refactor' into master

kyranet commented on commands/system/help.js:
```diff
@@ -1,6 +1,8 @@
 const { MessageEmbed } = require("discord.js");
+const emojis = ["⏮", "◀", "⏹", "▶", "⏭", "🔢"];
 const perpage = 10;
+let on = false;
```
This is a no-no. It should be in the reaction handler. The presence of this here makes the command usable only by one user **globally**. Should be restricted by channel or user.

YoussefExplicit replied:
✔️

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+      usage: "help <category/command/setting>",
+      aliases: ["h", "halp", "commands"],
     });
+    this.pages = async (message, msg2, page, sorted, type, level, r, this2, hm) => {
```
I'd say... too many parameters. It's unnecesary. Also, undescriptive names.

Also, this code is extremely duplicated with the command run method, and should be a method, not a property in the constructor.

YoussefExplicit replied:
It requires all these parameters, I've removed this2 and also given the names a bit more descriptiveness

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
     });
+    this.pages = async (message, msg2, page, sorted, type, level, r, this2, hm) => {
+      let n = 0;
+      sorted.forEach((c) => {
```
I still don't understand this, you could lowercase type once, and use [`Collection#reduce`](https://discord.js.org/#/docs/main/master/class/Collection?scrollTo=reduce) instead.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+
+      let num = 0;
+      let output = "";
+      const pg = Number(page) && Number(page) <= Math.ceil(n / perpage) ? Number(page) : 1;
```
As I said in the other PR, don't parse string to number +3 times, parse it once. Also, this line needs documentation.

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+      const pg = Number(page) && Number(page) <= Math.ceil(n / perpage) ? Number(page) : 1;
+      for (const c of sorted.values()) {
+        if (c.help.category.toLowerCase() === type.toLowerCase()) {
+          if (c.help.cat === "NSFW" && !message.channel.nsfw) return;
```
You know `return` doesn't make the loop skip the current iteration but stop it entirely with the function? You probably meant `continue;`

AndrewJWin replied:
If I recall correctly, that was the point of that one there. It's one I put in a while ago to prevent NSFW categories to be posted in non-nsfw channels.

kyranet replied:
Still, should use `continue;` so it shows the non-nsfw commands. Also, it's already filtered before:

https://github.com/YoussefElshemi/Misaki/blob/f3cb93b72f33315e3ad803f724b4277ba077039f/commands/system/help.js#L70

AndrewJWin replied:
If the category is NSFW, all the commands in that category **will** be NSFW...

kyranet replied:
Then check `if (type.toLowerCase() === "nsfw")` is true, beforehand, instead of doing a lot of heavy processing and finally, check it.

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+      let num = 0;
+      let output = "";
+      const pg = Number(page) && Number(page) <= Math.ceil(n / perpage) ? Number(page) : 1;
+      for (const c of sorted.values()) {
```
Instead of the code above with a forEach, you could iterate the commands only once with this for-of loop. Also, this needs a more descriptive name.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+        if (c.help.category.toLowerCase() === type.toLowerCase()) {
+          if (c.help.cat === "NSFW" && !message.channel.nsfw) return;
+          if (num < perpage * pg && num > perpage * pg - (perpage + 1)) {
+            if (level < this2.client.levelCache[c.conf.permLevel]) return;
```
What even is `this2`? The two lines above need documentation.

YoussefExplicit replied:
this2 is gone now

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+          if (c.help.cat === "NSFW" && !message.channel.nsfw) return;
+          if (num < perpage * pg && num > perpage * pg - (perpage + 1)) {
+            if (level < this2.client.levelCache[c.conf.permLevel]) return;
+            output = `${output}\n\`${message.settings.prefix + c.help.name}\` | ${c.help.description.length > 80 ? `${c.help.description.slice(0, 80)}...` : c.help.description}`;
```
Just use `+=` instead of using an anti-pattern.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+        .setDescription(`A list of commands in the ${type} category.\n(Total of ${num} commands in this category)\n\nTo get help on a specific command do \`${message.settings.prefix}help <command>\``)
+        .addField("Commands", output)
+        .setColor(message.guild.me.roles.highest.color || 5198940)
+        .setFooter(`Requested by ${message.author.tag}`, message.author.avatarURL());
```
Should use `displayAvatarURL` instead.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
 
-  async run(message, [type, page], level) {
+  async run(message, [type], level) {
+    let page = 1;
```
The variable declarations could be inlined.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
       .setTimestamp()
       .setColor(message.guild.me.roles.highest.color || 5198940)
-      .setFooter(`Requested by ${message.author.tag}`, message.author.avatarURL()); 
+      .setFooter(`Requested by ${message.author.tag}`, message.author.avatarURL());
```
Same as above, `displayAvatarURL`.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
-    const sorted = this.client.commands.sort((p, c) => p.help.category > c.help.category ? 1 :  p.help.name > c.help.name && p.help.category === c.help.category ? 1 : -1 );
+    const currentCategory = "";
+    const last = [];
+    const sorted = this.client.commands.sort((p, c) => (p.help.cat > c.help.category ? 1 : p.help.name > c.help.name && p.help.cat === c.help.category ? 1 : -1));
```
I'd highly suggest to sort the commands after filtering them.

YoussefExplicit replied:
sort them alphabetically?

kyranet replied:
No, you're sorting -> filtering -> filtering -> map -> filtering.
Should be: filter (all in one) -> sorting -> map
Or more efficient: filter+map (all in one, in the same loop) -> sorting

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
       const description = `Command category list\n\nUse \`${message.settings.prefix}help <category>\` to find commands for a specific category`;
-      const output = sorted.filter(c => !(level < 10 && c.help.category == "Owner") || !(c.help.category === "NSFW" && !message.channel.nsfw)).map(c => {
-        const cat = c.help.category.toProperCase();
+      const output = sorted.filter(c => !(level < 10 && c.help.category === "Owner") || !(c.help.category === "NSFW" && !message.channel.nsfw)).map((c) => {
```
I'd suggest:

Filter the commands with a for-of by the category name + this check. Once it's fully filtered, sort it and send it to the reaction handler.

At this moment, you're sorting, filtering (removing some of the sorted elements), filtering again (removing even more sorted elements). This becomes highly unefficient if a user types a category no command belongs to, and you'd still sort, filter, re-filter, and finally show nothing.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
     } else {
-      let n = 0;
-      sorted.forEach(c => {
+      sorted.forEach((c) => {
```
Same as above. Also, this code is duplicated.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
       let output = "";
-      let num = 0;
-      const pg = parseInt(page) && parseInt(page) <= Math.ceil(n / perpage) ? parseInt(page) : 1;
+      const pg = Number(page) && Number(page) <= Math.ceil(n / perpage) ? Number(page) : 1;
```
Duplicated code with the reaction handler. Same with all below.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+      await message.channel.send("I don\t have permission to remove reactions, please do this manually.");
+    }
+
+    if (msg2.embeds[0].title) {
```
You can put both checks together instead of two `if`s.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+    }
+    const msg2 = await message.channel.send(embed);
+    const totalpages = Math.ceil(num / perpage);
+    if (!message.guild.me.hasPermission(["MANAGE_MESSAGES"])) {
```
Don't you think you have processed thousands of operations (QuickSort, used by V8 to sort, is very expensive to run), and finally, you're checking for something that literally uses a couple of operations and that is very important for the entire thing to run.

YoussefExplicit replied:
It just notifies the user that they will have to remove reactions manually.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
 
-    return message.channel.send(embed);
+    const select = await msg2.createReactionCollector(
+      (reaction, user) => emojis.includes(reaction.emoji.name) && user.id === message.author.id,
```
It's much faster to check if both strings are equal than checking if an entry is in an array. Swap them.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
-        .addField("Command aliases", cm.conf.aliases.length == 0 ? "None" : cm.conf.aliases.join(", ") );
+        .addField("Command aliases", cm.conf.aliases.length === 0 ? "None" : cm.conf.aliases.join(", "));
+    }
+    const msg2 = await message.channel.send(embed);
```
This is an undescriptive name. Please rename this constant to another that represents better what it does.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
         .addField("Command description", cm.help.description)
         .addField("Command usage", `\`${cm.help.usage}\``)
-        .addField("Command aliases", cm.conf.aliases.length == 0 ? "None" : cm.conf.aliases.join(", ") );
+        .addField("Command aliases", cm.conf.aliases.length === 0 ? "None" : cm.conf.aliases.join(", "));
```
The `=== 0` is unneeded.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+      { time: 300000 },
+    );
+
+    if (select) {
```
Useless check.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
-    if (!embed.fields[0]) return;
 
-    return message.channel.send(embed);
+    const select = await msg2.createReactionCollector(
```
[`Message#createReactionCollector()`](https://discord.js.org/#/docs/main/stable/class/Message?scrollTo=createReactionCollector) does not return a Promise, but a [`ReactionCollector`](https://discord.js.org/#/docs/main/stable/class/ReactionCollector) instance.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+      select.on("collect", async (r) => {
+        switch (r.emoji.name) {
+          case "▶":
+            this.pages(message, msg2, Number(msg2.embeds[0].title.split(" ")[1].split("/")[0]) + 1, sorted, type, level, r, this, "forward");
```
Passing `this` to a function that is literally in the constructor is useless, anything in the constructor has access to the constructor.

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+            this.pages(message, msg2, Number(msg2.embeds[0].title.split(" ")[1].split("/")[0]) + 1, sorted, type, level, r, this, "forward");
+            break;
+          case "◀":
+            this.pages(message, msg2, Number(msg2.embeds[0].title.split(" ")[1].split("/")[0]) - 1, sorted, type, level, r, this, "backward");
```
You could cache the number somewhere. (To be honest, you have 3 undescriptive variables that hold integers, probably one of them represents this number)

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+              max: 1,
+              time: 300000,
+              errors: ["time"],
+            }).then(async (collected) => {
```
Useless Promise#then.

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+          embed.setTitle(type.toProperCase())
+            .setDescription(`A list of commands in the ${type} category.\n(Total of ${num} commands in this category)\n\nTo get help on a specific command do \`${message.settings.prefix}help <command>\``)
+            .addField("Commands", output)
+          :
```
This ternary isn't consistent. If `?` goes at the end of a line, then this one should aswell.

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
-        embed.setTitle("Command category help")
-          .setDescription(`A list of commands in the ${type} category.\n(Total of ${num} commands in this category)\n\nTo get help on a specific command do \`${message.settings.prefix}help <command>\`\n\n${num > 10 && pg === 1 ? `To view more commands do\` ${message.settings.prefix}help <category> 2\`` : "" }`)
-          .addField("Commands", output);
+        Math.ceil(num / perpage) === 1 ?
```
This ternary here is unneeded, the next 3-6 lines are duplicated. What is changed by this condition is the title.

YoussefExplicit replied:
✔️

YoussefExplicit replied:
✔️

kyranet commented on commands/system/help.js:
```diff
@@ -9,74 +11,170 @@ class Help extends Command {
+        .addField("Command aliases", cm.conf.aliases.length === 0 ? "None" : cm.conf.aliases.join(", "));
+    }
+    const msg2 = await message.channel.send(embed);
+    const totalpages = Math.ceil(num / perpage);
```
This is used twice before. You could avoid extra processing by declaring this before and use this instead.

kyranet commented on commands/system/help.js:
```diff
@@ -1,6 +1,8 @@
 const Command = require(`${process.cwd()}/base/Command.js`);
 const { MessageEmbed } = require("discord.js");
+const emojis = ["⏮", "◀", "⏹", "▶", "⏭", "🔢"];
```
It's a constant that never changes -> The name could be uppercase.

YoussefExplicit replied:
as in const Emojis?

kyranet replied:
No, as in, `EMOJIS`. The common convention is to use all-uppercase letters.

kyranet reviewed (changes requested):
And... the actual requested changes. There are 29 (basically in almost every 2-4 lines).

*Please DO NOT create any other PR, get your issues solved, but please, don't make another PR, it takes a while to review.*

YoussefExplicit committed:
hope some of this stuff is fixed.

YoussefExplicit committed:
oof

YoussefExplicit committed:
pls travis

YoussefExplicit committed:
last one ffs

kyranet commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
 
-  async run(message, [type, page], level) {
+    this.pages = async (message, helpmessage, pagenumber, sorted, type, level, reactions, direction) => {
+      let n = 0; // eslint-disable-line no-unused-vars
```
If you're not using this, don't use this.

AndrewJWin replied:
For some reason, I noticed this too while I was toying around with his version. It *is* required, but if I were to remove it, parts want to break..

kyranet commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
-  }
 
-  async run(message, [type, page], level) {
+    this.pages = async (message, helpmessage, pagenumber, sorted, type, level, reactions, direction) => {
```
Again, this should be a method in the class.

MrJacz replied:
Also what Kyra said, The method should be a class method and not in the constructor...

kyranet commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+      let output = "";
+      const pg = Number(pagenumber);
+      for (const c of sorted.values()) {
+        if (c.help.category.toLowerCase() === type.toLowerCase()) {
```
Again, you could do `type = type.toLowerCase()` to avoid lowercasing it in every single command.

kyranet commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+      let num = 0;
+      let output = "";
+      const pg = Number(pagenumber);
+      for (const c of sorted.values()) {
```
Undescriptive name. I'd suggest the name `cmd`.

kyranet commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
       const description = `Command category list\n\nUse \`${message.settings.prefix}help <category>\` to find commands for a specific category`;
-      const output = sorted.filter(c => !(level < 10 && c.help.category == "Owner") || !(c.help.category === "NSFW" && !message.channel.nsfw)).map(c => {
-        const cat = c.help.category.toProperCase();
+      const output = sorted.map((c) => {
```
I requested a single for loop that does the entire filtering and mapping in the valid keys, then, sorting the values. Sorting, map, and finally, filter, is much slower than the previous setup.

kyranet reviewed (changes requested):
There's still a lot of code duplication between Help#pages and Help#run.

Ideally, you'd have three methods:

- `Help#run` handles the permissions and everything before running the reaction handler.
- `Help#process` processes the data given the query, this is done in both methods of this class at this moment, using this method could reduce the extreme complexity.
- `Help#pages` takes the collector and processes it, it does not filter, but solely handle the reactions.

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+    this.pages = async (message, helpmessage, pagenumber, sorted, type, level, reactions, direction) => {
+      let n = 0; // eslint-disable-line no-unused-vars
+      for (const c of sorted.values()) {
+        if (c.help.category.toLowerCase() === type.toLowerCase()) {
```
this can be a one-liner

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+        }
+      }
+      reactions.users.remove(message.author);
+      if (direction === "forward") {
```
this could be a one-liner.

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+      if (direction === "forward") {
+        if (pg > Math.ceil(num / perpage)) return;
+      }
+      if (direction === "backward") {
```
One-liner here as well

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
         if (currentCategory !== cat && !type) {
-          currentCategory = cat;
-          return `\n\`${message.settings.prefix}help ${cat.toLowerCase()}\` | Shows ${cat} commands`;
+          if (last.includes(cat.toLowerCase())) return;
```
Instead of `toLowerCase`ing it every time you use it, just do `const cat = c.help.category.toLowerCase();`
and add `.toLowerCase()` to the check below (to currentCategory)

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+
+    const select = msg2.createReactionCollector(
+      (reaction, user) => EMOJIS.includes(reaction.emoji.name) && user.id === message.author.id,
+      { time: 300000 },
```
Should be 30 seconds at max

YoussefExplicit replied:
30 seconds to use a help command??!?!!?

MrJacz replied:
It doesn't need 5 minutes, Not for a help command...

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+    let on = false;
+    select.on("collect", async (r) => {
+      switch (r.emoji.name) {
+        case "▶":
```
Make one variable for `Number(msg2.embeds[0].title.split(" ")[1].split("/")[0]) - 1`

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
+    });
+
+    select.on("end", (r, reason) => {
+      if (reason === "time") {
```
this can be a one-liner

MrJacz commented on commands/system/help.js:
```diff
@@ -1,83 +1,162 @@
 const Command = require(`${process.cwd()}/base/Command.js`);
 const { MessageEmbed } = require("discord.js");
+const EMOJIS = ["⏮", "◀", "⏹", "▶", "⏭", "🔢"];
```
This variable should not be in full caps `emojis` will do

fratik replied:
@MrJacz kyra said to do `EMOJIS`

MrJacz replied:
Why? (╯°□°）╯︵ ┻━┻

tech6hutch replied:
It's a convention for constants. I guess most things here are technically constants, though...

MrJacz reviewed (changes requested):
Leaving some requested changes, These are purely to clean the command up and some other nit picks

YoussefExplicit committed:
Hope this cleared some stuff.

fratik commented:
```
/home/travis/build/NotAWeebDev/Misaki/commands/system/help.js
  155:30  error  Unexpected space before function parentheses  space-before-function-paren
  158:2   error  Missing semicolon                             semi
```

YoussefExplicit committed:
Fixed requested things for jacz.

YoussefExplicit committed:
Update help.js

YoussefExplicit committed:
30 seconds help

YoussefElshemi commented:
All fixed 👍

YoussefExplicit committed:
this should be it :P

YoussefExplicit committed:
Forgot to check if the user is owner

Forgot to check if the user is the owner, and if so, it'll remove any owner commands.

goatwithakb commented on commands/system/help.js:
```diff
@@ -9,75 +10,146 @@ class Help extends Command {
+        if (cmd.help.category === "NSFW" && !message.channel.nsfw) continue;
+        if (num < perpage * pg && num > perpage * pg - (perpage + 1)) {
+          if (level < this.client.levelCache[cmd.conf.permLevel]) continue;
+          output += `\n\`${message.settings.prefix + cmd.help.name}\` | ${cmd.help.description.length > 80 ? `${cmd.help.description.slice(0, 80)}...` : cmd.help.description}`;
```
Am I right in saying that you reconstruct the embed for each page, every time someone cycles forward/backward?

goatwithakb replied:
(why did this show up as a review >.<)

YoussefExplicit replied:
the answer is yes

goatwithakb commented on commands/system/help.js:
```diff
@@ -9,75 +10,146 @@ class Help extends Command {
+            time: 300000,
+            errors: ["time"]
+          });
+          const pagenumber = Number(whichpage.first().content);
```
Do you check if the pagenumber is inside the bounds of how many pages there are (will it crash if a user enters -1)

AndrewJWin replied:
No, there is a check that verifies that the number is equal to or greater to one, otherwise it defaults to 1

kyranet reviewed (changes requested):
For 7th time already. The `pages` method **MUST ONLY** handle the reaction handler. The `run` method must only parse and prepare the input, and finally, a third but also optional method called `prepare` to reduce the complexity from both methods.

Right now, the complexity of both methods are probably over 40, which makes them incredibly hard to read, follow, and understand. I will not approve this PR until each method do exclusively what they are supposed to do.

Also, the method `pages` still has a lot of arguments, they could be reduced.

YoussefElshemi commented:
@kyranet First of all, I don't intend any rudeness by saying this. I don't understand why complexity is important, Misaki doesn't support self-hosting so a user doesn't need to understand. I don't see the problem with the function as of now, everything works fast and I've done everything you've asked for, I believe that these requested changes are deemed unnecessary and I think it should be approved. If you don't wish to, that is fine, but I've spent too much time perfecting something that works as intended. From this point on, anyone is free to re-attempt/re-use the code and modify it to fulfill kyra's needs. Thanks.

YoussefElshemi commented:
Take that back, might try to reattempt it :P

DevYukine commented on commands/system/help.js:
```diff
@@ -1,5 +1,6 @@
 const Command = require(`${process.cwd()}/base/Command.js`);
 const { MessageEmbed } = require("discord.js");
+const EMOJIS = ["⏮", "◀", "⏹", "▶", "⏭", "🔢"];
```
could be moved to the constructor tbh since you use it inside the class

DevYukine commented on commands/system/help.js:
```diff
@@ -9,75 +10,146 @@ class Help extends Command {
 
-module.exports = Help;
\ No newline at end of file
+Array.prototype.unique = function() {
```
**Big nope** from me to this, this is bad practise at all but if you actuall need to do that, add this inside the function.js file atleast.
